shoop/core/migrations/0013_address_changes.py

```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+import django_countries.fields
+import shoop.core.utils.name_mixin
+import django.db.models.deletion
+import shoop.core.models._base
+import shoop.core.fields
+import shoop.utils.models
+
+
+def from_data(cls, data, immutable=False):
+    """
+    Create Address from a data dictionary.
+
+    :type data: dict[str,str]
+    :type immutable: boolean
+    :rtype: Unsaved Address
+    """
+    if immutable:
+        return cls.objects.get_or_create(**data)[0]
+    return cls(**data)
+
+
+def populate_immutable_addresses(cls, field, address_cls, address, address_data):
+    """
+    Update existing order address with immutable address which is got
+    Address from_data. This makes sure that only one identical immutable
+    address is created.
+
+    :param cls: Order class
+    :param field: field to update
+    :param address_cls: Address class
+    :param address: original address to be updated
+    :param address_data: address data
+    :return: None
+    """
+    query = {field: address}
+    objects = cls.objects.filter(**query)
+    if objects.exists():
+        immutable_address = from_data(address_cls, address_data, immutable=True)
+        immutable_address.save()
+        updated_field = {field: immutable_address}
+        objects.update(**updated_field)
+
+
+def copy_addresses(apps, schema_editor):
+    """
+    Copy existing addresses to new Mutable and Immutable addresses
+    For every address new MutableAddress is created with same id
```

Comment:
@akx is right. The autoincrement should be updated. Though I don't think there is any way to do that in a DBMS agnostic way...

This could be a big problem. Maybe the immutable addresses should be created without manually assigning any ids and just store a mapping "order_id -> immutable_address_id". This would mean that address ids in order has to be NULLed while creating the mapping and then finally when the foreign keys of order table are updated, their values should be set from the mapping.